### PR TITLE
Migrate phoenix gen to igniter

### PR DESCRIPTION
**Pull request by srikanthkyatham**
### Contributor checklist

Atempt 1 to create PR. Once direction is right. I need to figure out how create tests. Perhaps embed a test project to generate the live views from it. 

- [ ] Bug fixes include regression tests
- [x] Features include unit/acceptance tests

### CODE REVIEW

File: lib/ash_phoenix/gen/live.ex [hunk 1]
```diff
@@ -137,7 +151,10 @@ defmodule AshPhoenix.Gen.Live do
       |> EEx.eval_file(assigns: assigns)
       |> formatter_function.()
 
-    Mix.Generator.create_file(destination_path, contents, generate_opts)
+    # igniter
+    Igniter.create_new_file(igniter, destination_path, contents, generate_opts)
+
+    # Mix.Generator.create_file(destination_path, contents, generate_opts)
```
Comment (zachdaniel) [inline lib/ash_phoenix/gen/live.ex:157]: Extra comments.

**Comment by zachdaniel** [comment]
This looks great! TO test it, see how we do it in the igniter tests. You'd define a resource in `test/support` to use as the basis, and then do something like this:

```elixir
import Igniter.Test

# in your test
test_project()
|> Igniter.compose_task("ash_phoenix.gen.live", [....])
|> assert_creates("....")
```

One challenge will be a dealing with the mix shell inputs, but there are tools for doing that 
https://hexdocs.pm/mix/main/Mix.Shell.Process.html

IIRC you can send the answers to the prompts in order using  this kind of code 
`send(self(), {:mix_shell_input, :prompt, "Pretty cool"})`
and then when you run the task it will use those inputs.

One other thing worth noting is that there is an active PR to set up a *phoenix* project instead of just `test_project()`, in this PR:

https://github.com/ash-project/igniter/pull/109

We can either wait on that or proceed without it. We can probably proceed without it.

**Comment by srikanthkyatham** [comment]
Cool I will write the tests and make the changes you pointed out. Yeah probably we can continue with out the other PR. Once it is merged I would be happy to contribute back.

**Comment by zachdaniel** [comment]
Great work, thank you!

**Comment by srikanthkyatham** [comment]
@zachdaniel 

I figured the Mix shell prompts. Now I am stuck with the test_project setup

`
    igniter =
      assert test_project(
               files: %{
                 "lib/test/support.ex" => """
                 defmodule Test.Support do
                   use Ash.Domain

                   resources do
                     resource Test.Support.Ticket
                   end
                 end
                 """)
                 |> apply_igniter!()
                 |> |> Igniter.compose_task("ash_phoenix.gen.live", [])
`

In the task, it is failing at

`
if !Spark.Dsl.is?(domain, Ash.Domain) do
`
It looks like I am missing sth. What are those is not quite evident. The tried dbg() the outputs. Not very helpful. Any pointers would be helpful.

**Comment by zachdaniel** [comment]
Right. So, the difficulty here is that those files are not compiled, and this task uses introspection to determine what to do.

This is a pretty hard problem TBH. I think we could add something like this to igniter (probably in test to start)

```elixir
def compile(igniter) do
  igniter.rewrite
  |> Stream.map(fn source -> 
     Kernel.ParallelCompiler.async(fn -> 
       Code.compile_string(Rewrite.Source.get(source, :content))
     end)
  end)  
end
```

**Comment by srikanthkyatham** [comment]
Hi @zachdaniel 

One other issue, I tried your compile function. It seems that 
```elixir
:erlang.get(:elixir_compiler_info)
```
is returning undefined. I have not played around Kernel.ParallelCompiler that much. Do you have any insight in this.

**Comment by zachdaniel** [comment]
🤔 yeah, so that may not work. I think for now, for these tests, you will have to test with actually defined domains & resources. Like in the `test/support` folder.

**Comment by srikanthkyatham** [comment]
Hi

Ok I will try to create the resources in the `test/support` folder and continue with the tests.

**Comment by srikanthkyatham** [comment]
Hi @zachdaniel 

Almost got it working. The last battle is with the indentation in the assert_creates

Screenshot 2024-11-05 at 22 02 31

I have tried 
1. Mix.Tasks.Format.formatter_for_file
2. Code.format_string!
3. String split on the code. I found you have used in one your tests. 

The above is what is blocker now.

**Comment by srikanthkyatham** [comment]
Adding LiveView.Formatter helped a bit. Now there is a space at the end of file created

Screenshot 2024-11-05 at 22 39 15

**Comment by srikanthkyatham** [comment]
All live gen test pass now. 

Screenshot 2024-11-06 at 9 44 04

**Comment by zachdaniel** [comment]
Amazing 🤩 I will give this a try this week and we can get it merged! 🥳

**Review by zachdaniel — APPROVED** [review]
LGTM! Mind rebasing and I'll merge?

**Comment by srikanthkyatham** [comment]
I will do it today in few hours

**Comment by srikanthkyatham** [comment]
Rebased the pr with main.

**Comment by zachdaniel** [comment]
🚀 Thank you for your contribution! 🚀

**Comment by zachdaniel** [comment]
Great work 🙇

**Pull request merged**